File: dominoes-web-app/src/lib/shaper-domino/pageSetup.ts
import type { Measurement } from './measurement';
import { inches, millimeters } from './measurementHelpers';

export type PageSetup = {
	width: Measurement;
	height?: Measurement;
	margin?: Measurement;
};

export const PageSetup = {
	letterPortrait: (margin?: Measurement): PageSetup => ({
		width: inches(8.5),
		height: inches(11),
		margin: margin ?? inches(0.25)
	}),
	legalPortrait: (margin?: Measurement): PageSetup => ({
		width: inches(8.5),
		height: inches(14),
		margin: margin ?? inches(0.25)
	}),
	letterLandscape: (margin?: Measurement): PageSetup => ({
		width: inches(11),
		height: inches(8.5),
		margin: margin ?? inches(0.25)
	}),
	A4Portrait: (margin?: Measurement): PageSetup => ({
		width: millimeters(210),
		height: millimeters(297),
		margin: margin ?? millimeters(10)
	}),
	A3Portrait: (margin?: Measurement): PageSetup => ({
		width: millimeters(297),
		height: millimeters(420),
		margin: margin ?? millimeters(10)
	}),
	A4Landscape: (margin?: Measurement): PageSetup => ({
		width: millimeters(297),
		height: millimeters(210),
		margin: margin ?? millimeters(10)
	}),
	continuousLabel: (width: Measurement, margin?: Measurement): PageSetup => ({
		width,
		height: undefined,
		margin: margin ?? inches(0.125)
	})
};

File: dotnet/DominoFactory/PageSetup.cs
// <copyright file="PageSetup.cs">Copyright (c) Peter Rosser. All rights reserved.</copyright>

namespace ShaperDominoLib;

public readonly record struct PageSetup(Measurement Width, Measurement? Height, Measurement? Margin)
{
    public static PageSetup LetterPortrait(Measurement? margin = null) => new(8.5.Inches(), 11.Inches(), margin ?? 0.25.Inches());
    public static PageSetup LetterLandscape(Measurement? margin = null) => new(11.Inches(), 8.5.Inches(), margin ?? 0.25.Inches());
    public static PageSetup A4Portrait(Measurement? margin = null) => new(210.Millimeters(), 297.Millimeters(), margin ?? 10.Millimeters());
    public static PageSetup A4Landscape(Measurement? margin = null) => new(297.Millimeters(), 210.Millimeters(), margin ?? 10.Millimeters());

    public static PageSetup ContinuousLabel(Measurement width, Measurement? margin = null) => new(width, null, margin ?? 0.125.Inches());
}